quic/s2n-quic-platform/src/io/testing/model.rs

```diff
@@ -0,0 +1,303 @@
+    let value = value * u64::MAX as f64;
+    value.round() as u64
+}
+
+struct State {
+    delay: AtomicU64,
+    jitter: AtomicU64,
+    network_jitter: AtomicU64,
+    transmit_rate: AtomicU64,
+    retransmit_rate: AtomicU64,
+    corrupt_rate: AtomicU64,
+    drop_rate: AtomicU64,
+    mtu: AtomicU16,
+    max_inflight: AtomicU64,
+    current_inflight: AtomicU64,
+}
+
+impl Default for State {
+    fn default() -> Self {
+        Self {
+            delay: AtomicU64::new(Duration::from_millis(50).as_micros() as _),
+            jitter: AtomicU64::new(0),
+            network_jitter: AtomicU64::new(0),
+            transmit_rate: AtomicU64::new(u64::MAX),
+            retransmit_rate: AtomicU64::new(0),
+            corrupt_rate: AtomicU64::new(0),
+            drop_rate: AtomicU64::new(0),
+            mtu: AtomicU16::new(MaxMtu::default().into()),
+            max_inflight: AtomicU64::new(u64::MAX),
+            current_inflight: AtomicU64::new(0),
+        }
+    }
+}
+
+impl Network for Model {
+    fn execute(&mut self, buffers: &Buffers) -> usize {
+        let jitter = self.jitter();
+        let network_jitter = self.network_jitter();
+        let transmit_rate = self.transmit_rate();
+        let retransmit_rate = self.retransmit_rate();
+        let corrupt_rate = self.corrupt_rate();
+        let drop_rate = self.drop_rate();
+        let mtu = self.mtu() as usize;
+
+        let now = super::time::now();
+        let mut transmit_time = now + self.delay();
+        let transmit_time = &mut transmit_time;
+
+        let mut transmit = |packet: Cow<Packet>| {
+            // drop the packet if it's over the current MTU
+            if packet.payload.len() > mtu {
+                return 0;
+            }
+
+            // drop the packet if enabled
+            if drop_rate > 0 && super::rand::gen::<u64>() < drop_rate {
+                return 0;
+            }
+
+            let mut packet = packet.into_owned();
+
+            if corrupt_rate > 0 && super::rand::gen::<u64>() < corrupt_rate {
+                // randomly truncate the payload
+                let num_bytes = super::rand::gen_range(0..packet.payload.len());
+                if num_bytes > 0 {
+                    packet.payload.truncate(num_bytes);
+                }
+
+                // randomly swap bytes in the payload
+                let num_bytes = super::rand::gen_range(0..packet.payload.len());
+                if num_bytes > 0 {
+                    super::rand::swap_count(&mut packet.payload, num_bytes);
+                }
+
+                // randomly rewrite bytes in the payload
+                let num_bytes = super::rand::gen_range(0..packet.payload.len());
+                if num_bytes > 0 {
+                    for _ in 0..num_bytes {
+                        let index = super::rand::gen_range(0..packet.payload.len());
+                        packet.payload[index] = super::rand::gen();
+                    }
+                }
```

Comment:
Hmm not sure. Since everything in quic is authenticated any one of those things should cause an authentication failure. Were you thinking of having separate knobs for each type of mutation?

Reply:
I guess I was wondering why go through the effort of corrupting the packet in all these ways?

Reply:
mostly to see if we have any panics/errors in the authentication mechanism. but i can make it simpler if you would prefer.

Reply:
I was thinking if you are corrupting in all 3 ways all the time it might be missing out on some bug that was only triggered by one of them. But if the `rand::gen_range` returns a 0 randomly it would skip some of mutations, so I guess you have those cases covered too. I'm fine to leave it

Reply:
Yep that's right; it'll do some random combination of the 3.